pkg/addonmanager/controllers/agentdeploy/controller.go

```diff
@@ -131,52 +131,86 @@ func (c *addonDeployController) sync(ctx context.Context, syncCtx factory.SyncCo
 	}
 
 	managedClusterAddon, err := c.managedClusterAddonLister.ManagedClusterAddOns(clusterName).Get(addonName)
-	if errors.IsNotFound(err) {
-		c.cache.removeCache(addonName, clusterName)
-		return nil
-	}
 	if err != nil {
+		if errors.IsNotFound(err) {
+			c.cache.removeCache(addonName, clusterName)
+			return nil
+		}
 		return err
 	}
-
-	if !managedClusterAddon.DeletionTimestamp.IsZero() {
-		c.cache.removeCache(addonName, clusterName)
-		return nil
-	}
-
+	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
 	owner := metav1.NewControllerRef(managedClusterAddon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
 
-	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
 	objects, err := agentAddon.Manifests(managedCluster, managedClusterAddon)
 	if err != nil {
 		return err
 	}
 	if len(objects) == 0 {
 		return nil
 	}
-
-	work, err := buildManifestWorkFromObject(clusterName, addonName, objects)
+	deployWork, hookWork, err := buildManifestWorkFromObject(clusterName, addonName, objects)
 	if err != nil {
 		return err
 	}
-	work.OwnerReferences = []metav1.OwnerReference{*owner}
+
+	if hookWork != nil {
+		if !hasFinalizer(managedClusterAddonCopy.GetFinalizers(), constants.PreDeleteHookFinalizer) {
+			managedClusterAddonCopy.Finalizers = append(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+			_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+				ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	if !managedClusterAddon.DeletionTimestamp.IsZero() {
+		c.cache.removeCache(addonName, clusterName)
```

Comment:
should hookWork also be put in the cache?

Reply:
handle following the deploywork